core/services/capreg/types.go

```diff
@@ -0,0 +1,148 @@
+type NodeOperator struct {
+	// ID is the unique identifier of the Node Operator in the CR.
+	ID *big.Int
+
+	// Admin is the address of the admin of the Node Operator.
+	Admin string
+
+	// Name is the name of the Node Operator.
+	Name string
+}
+
+type Node struct {
+	NodeOperatorID *big.Int
+	// ID is the unique identifier of the Node in the CR,
+	// which is a RageP2P ID.
+	ID []byte
+
+	// CapabilityIDs are the unique identifiers of the capabilities of this Node.
+	CapabilityIDs []CapabilityID
+}
+
+type ResponseType int
+
+const (
+	ResponseTypeReport ResponseType = iota
+	ResponseTypeObservationIdentical
+)
+
+// Capability represents a capability in the Capability Registry.
+// These can be thought of roughly as "things DONs can do".
+type Capability struct {
+	// Type is the type of the capability, e.g trigger, action, consensus, etc.
+	Type string
+
+	// Version is the version of the capability, in semantic versioning notation, e.g 1.2.3.
+	Version string
+
+	// ID is the unique identifier of the capability in the CR.
+	// It is calculated as keccak256(abi.encode(capabilityType, capabilityVersion)).
+	ID CapabilityID
+
+	// ResponseType indicates whether remote response requires aggregation
+	// or is an OCR report. There are multiple ways to aggregate.
+	ResponseType ResponseType
+
+	// ConfigurationContractAddress is the address of the configuration contract for this capability.
+	ConfigurationContractAddress string
+}
+
+// State mirrors the state in the onchain capability registry.
+type State struct {
+	Capabilities   map[CapabilityID]Capability
+	CapabilityIDs  []CapabilityID
+	DONs           []DON
+	CapabilityDONs map[string][]DON
+}
+
+// Local is an interface for updating local capability registry state.
+// Implementations of this can choose to filter out capabilities or DONS
+// as they see fit.
+// For example, for a capability router, it would need to know all DONs that
+// support a particular capability.
+// However, for a local capability launcher, it may only need to know about
+// capabilities and DONs that the local node participates in.
+type Local interface {
+	// Sync is called when the onchain state of the capability registry is fetched.
+	// It is up to the implementation to decide how to handle the new state.
+	Sync(ctx context.Context, s State) error
+
+	// Close is called when the local registry is shutting down.
+	Close() error
+}
+
+// CapabilityService is an interface that must be implemented by capabilities that get started
+// via the capability registry synchronization process.
+//
+//go:generate mockery --name CapabilityService --output ./mocks/ --case underscore
+type CapabilityService interface {
+	// CapabilityID returns the unique identifier of the capability being implemented.
+	CapabilityID() CapabilityID
+
+	// Start is called on startup of the local registry or when a new DON is added
+	// to the capability registry while the local registry is running.
+	Start(ctx context.Context, d DON) error
```

Comment:
@makramkd You're might be missing some stuff here -- all capabilities will need to accept a standard set of core node APIs (eg. RelayerSet, PipelineRunner, etc) -- is this going to happen lower down?